### DISCUSSION TITLE
Out of source build error

### DISCUSSION THREAD

**Discussion by pamely** [robotology/community#654] [Q&A] [answered]
I am currently trying the "Compiling a YARP hello world program" available on the following l[ink ](ttps://www.yarp.it/latest//yarp_cmake_hello.html).  After running `cmake` to build the package hello, I got the error below: 

```plaintext
CMake Error at CMakeLists.txt:16 (find_package):
  By not providing "FindYARP.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "YARP", but
  CMake did not find one.

  Could not find a package configuration file provided by "YARP" with any of
  the following names:

    YARPConfig.cmake
    yarp-config.cmake

  Add the installation prefix of "YARP" to CMAKE_PREFIX_PATH or set
  "YARP_DIR" to a directory containing one of the above files.  If "YARP"
  provides a separate development package or SDK, be sure it has been
  installed.
```
While I understand that I have to prepend "YARP" to CMAKE_PREFIX_PATH, I don't know how to find the path to YARP installation prefix. 
Please, how can I fix this? 

As suggested in the tutorial, I have the following directory structure: 
`/yarp/hello/CMakeLists.txt`
`/yarp/hello/hello.cpp`
`/yarp/hello/build`

I am running `cmake ..` in the build folder. 

Thank you, 
Pamely

**Comment by pattacini** [comment, marked as answer]
How did you build `YARP`? 

## ⬜ Just build w/o install
If you didn't install it (you just typed `make`), then you ought to follow one of the proposed suggestions:

```console
  Add the installation prefix of "YARP" to CMAKE_PREFIX_PATH or set
  "YARP_DIR" to a directory containing one of the above files.  If "YARP"
  provides a separate development package or SDK, be sure it has been
  installed.
```

### ⚪ Suggestion 1 (`CMAKE_PREFIX_PATH`)
```console
cd /yarp/hello/build
cmake -DCMAKE_PREFIX_PATH=/path_to_yarp_build_dir ..
```
where `/path_to_yarp_build_dir` shall point to the directory where you built `YARP`.

Alternatively, you could export `CMAKE_PREFIX_PATH` as an env variable.

### ⚪ Suggestion 2 (`YARP_DIR`)
Put the line below into your `.bashrc` file
```console
export YARP_DIR=/path_to_yarp_build_dir
```

## ⬜ Build and install
By contrast, if you installed YARP by typing `make install`, just have a look at where the compilation products have been installed by inspecting the build log or checking the value of the `CMAKE_INSTALL_PREFIX` cmake variable.

Usually, `CMAKE_INSTALL_PREFIX` points to `/usr/local/` so that you should be able to let cmake find what's needed w/o fiddling any further.

If `CMAKE_INSTALL_PREFIX` points to a user-defined directory, then one needs to fill in other env variables like `PATH`, `LD_LIBRARY_PATH`, `YARP_DATA_DIRS` with the correct paths.
